dadi/lib/search/index.js

```diff
@@ -1,77 +1,607 @@
+ */
+Search.prototype.initialiseConnections = function () {
+  let searchConfig = config.get('search')
+
+  this.wordConnection = Connection(
+    {
+      collection: this.wordCollection,
+      database: searchConfig.database,
+      override: true
+    },
+    this.wordCollection,
+    searchConfig.datastore
+  )
+
+  this.searchConnection = Connection(
+    {
+      collection: this.searchCollection,
+      database: searchConfig.database,
+      override: true
+    },
+    this.searchCollection,
+    searchConfig.datastore
+  )
+
+  this.wordConnection.setMaxListeners(35)
+  this.searchConnection.setMaxListeners(35)
+}
+
+/**
+ * Apply Index Listeners
+ * Fires a call to the data controllers index method with the schemas index rules.
+ */
+ // TODO: this will change with @eduardo's Connection Recovery branch
+Search.prototype.applyIndexListeners = function () {
+  this.wordConnection.once('connect', database => {
+    database.index(this.wordCollection, this.getWordSchema().settings.index)
+  })
+
+  this.searchConnection.once('connect', database => {
+    database.index(this.searchCollection, this.getSearchSchema().settings.index)
+  })
+}
+
+/**
+ * Find documents in the "words" collection matching the specified searchTerm, using the results of the query
+ * to fetch results from the current collection's search collection, ultimately leading to a set of IDs for documents
+ * that contain the searchTerm
+ *
+ * @param {String} searchTerm - the search query passed to the collection search endpoint
+ * @return {Promise} - resolves with a query containing IDs of documents that contain the searchTerm
+ */
+Search.prototype.find = function (searchTerm) {
+  debug(this.canUse() ? 'search enabled' : 'search disabled')
+
+  debug('find in %s: %s', this.searchCollection, searchTerm)
+
+  let tokenized = this.analyser.tokenize(searchTerm)
+
+  return this.getWords(tokenized).then(words => {
+    let wordIds = words.results.map(word => word._id.toString())
+
+    debug('searching %s for words %o', this.searchCollection, words.results.map(word => word.word))
+
+    return this.searchConnection.datastore.search({
+      words: wordIds,
+      collection: this.searchCollection,
+      schema: this.getSearchSchema().fields,
+      settings: this.getSearchSchema().settings,
+      opions: { limit: pageLimit }
+    }).then(wordInstances => {
+      wordInstances = wordInstances.map(instance => instance._id.document)
+
+      return {
+        _id: {
+          '$containsAny': wordInstances
+        }
+      }
+    })
+  })
+}
+
+/**
+ * Removes entries in the collection's search collection that match the specified documents
+ * @param {Array} documents - an array of documents for which to remove word instances
+ * @return {Promise} - Query to delete instances with matching document ids.
+ */
+Search.prototype.delete = function (documents) {
+  if (!this.canUse()) {
+    return Promise.resolve()
+  }
+
+  if (!Array.isArray(documents)) {
```

Comment:
Could we merge this with the first `if`, perhaps?

Reply:
ok